### DISCUSSION TITLE
Using `TileLayer` to add WMS overlay

### DISCUSSION THREAD

**Discussion by tbergama** [visgl/deck.gl#6885] [Q&A] [answered]
Hi,

I'm trying to overlay weather radar data from [NOAA's nowcoast WMS](https://nowcoast.noaa.gov/) over a basemap using a `TileLayer`. Unfortunately the nowcoast WMS does not provide XYZ indexing, it only supports passing a bounding box. So I am using the `getTileData` prop and using the bounding box from `tile.bbox` to construct URLs for each tile.

After I implemented this, I noticed that the tiles seemed to be drawn in the incorrect locations when the map is zoomed out, but as the map is zoomed in the tiles moved with respect to the base map and start approaching their expected location. I found [a discussion which observed similar behavior](https://github.com/visgl/deck.gl/discussions/5396#discussion-2227748) and realized that I was specifying `EPSG:4326` as the bbox CRS in my URLs, and as @kylebarron mentions in that thread - "`TileLayer` only works with `EPSG:3857`". The URL will also except `EPSG:3857` as the bbox CRS, but when I switch the bbox CRS in the URLs, the WMS returns a bunch of empty PNGs. After looking at the data inside `tile.bbox`, I noticed that it's units look like degrees but according to the nowcoast docs and [this](https://epsg.io/3857), `EPSG:3857` uses meters.

Does the `tile.bbox` object I'm accessing in `getTileData` have a CRS of `EPSG:3857`? If so, do they have units of degrees or meters? Will I need to convert degrees to meters before constructing the URL, or try something else entirely? Any guidance on getting a TileLayer to display data from this WMS is very much appreciated. Thanks!

**Comment by kylebarron** [comment, marked as answer]
I think there was an example of WMS somewhere. You might be able to search back in issues and find one.

The TileLayer uses the OSM indexing system, so the tiles are using web mercator, not global geodetic projection. But it isn't technically EPSG:3857 because that uses mercator meters as the unit of measurement. I believe the bbox arg you get in the TileLayer uses EPSG:4326 to describe the web mercator tile.

**Reply by tbergama** [reply]
Got it! Could have sworn that I had looked through all the issues, but I guess I missed [this one](https://github.com/visgl/deck.gl/discussions/6450) that along with your explanation here helped me figure it out. Thank you!

The [nowcoast ArcGIS REST Map Service](https://nowcoast.noaa.gov/help/#!section=rest-usage) (which is distinguished from the WMS in their docs) allows you to specify the bounding box CRS and the output image CRS separately. By specifying `EPSG:4326` as the bounding box CRS and `EPSG:3857` as the output image CRS I was able to get the data to display in the correct location and stop moving around when zooming out.

For anyone interested, here is an example URL:

`https://nowcoast.noaa.gov/arcgis/rest/services/nowcoast/radar_meteo_imagery_nexrad_time/MapServer/export?transparent=true&format=png8&layers=show:&bboxSR=4326&imageSR=3857&size=1024,1024&f=image&bbox=-180,0,-90,66.51326044311186`

_The key is specifying_ `bboxSR=4326` _and_ `imageSR=3857`

And here is my full TileLayer:

```
new TileLayer({
            id: "weather-overlay",
            getTileData: (tile) => {
              const { west, north, east, south } = tile.bbox;
              const base_url = "https://nowcoast.noaa.gov/arcgis/rest/services/nowcoast/radar_meteo_imagery_nexrad_time/MapServer/";
              const url_params = "export?transparent=true&format=png8&layers=show:&bboxSR=4326&imageSR=3857&size=1024,1024&f=image";
              const url_bbox = `&bbox=${west},${south},${east},${north}`;

              return base_url + url_params + url_bbox;
            },

            renderSubLayers: props => {
              const {
                bbox: { west, south, east, north }
              } = props.tile;

              return new BitmapLayer({
                id: nanoid(),
                image: props.data,
                opacity: 0.3,
                bounds: [west, south, east, north]
              })
            }
          })
```

**Reply by tbergama** [reply]
Also for anyone interested, I was able to get `TileLayer` to correctly display another WMS that _did not_ allow me to specify different CRS for bounding box and output image by using [proj4.js](https://www.npmjs.com/package/proj4) to transform the `tile.bbox` coords from `EPSG:4326` to `EPSG:3857` and then specifying `EPSG:3857` as the CRS in the URL. 

Transformation looks like this:
```
const [west, south] = proj4('EPSG:4326', 'EPSG:3857', [tile.bbox.west, tile.bbox.south]);
const [east, north] = proj4('EPSG:4326', 'EPSG:3857', [tile.bbox.east, tile.bbox.north]);
```

**Comment by Pessimistress** [comment]
I think this is a great use case for #6848. This feature will be available in v8.8. We should put together an example once an alpha version is ready.

**Reply by tbergama** [reply]
Interesting! While in this case it turned out that the service allows you to specify bounding box CRS and output image CRS separately, there are other services I am interested in using that do not. The tile shape would still be a square for these, but it would be helpful to be able to set CRS of `tile.bbox` to be `EPSG:3857` so that the bounding box can be given in the CRS that is needed for the output image CRS. Would this feature help with this?

**Reply by kylebarron** [reply]
If you still want to render data using the OSM tiling index, then the simplest solution would be to reproject the four `bbox` coords into `EPSG:3857` and use that for both the bbox coords and the image. You can do that with e.g. [`sphericalmercator.convert`](https://github.com/mapbox/sphericalmercator#convertbbox-to). Not sure but I think [`@math.gl/web-mercator`](https://math.gl/modules/web-mercator/docs/api-reference/web-mercator-utils) (already a dependency used by deck.gl) is essentially the same but you need an additional linear transformation to convert from the 512x512 Web Mercator to mercator meters.

**Reply by tbergama** [reply]
Nice, yeah I used [proj4.js](https://www.npmjs.com/package/proj4) to reproject the bbox coords to `EPSG:3857` and my `TileLayer` is working much better now. I'll check out [`@math.gl/web-mercator`](https://math.gl/modules/web-mercator/docs/api-reference/web-mercator-utils) though to see if I can remove a dependency. Thanks for the help!

**Reply by kylebarron** [reply]
That works too. But proj4.js is overkill if all you need is to go between EPSG:4326 and Web Mercator (colloquially "GOOGLE" https://epsg.io/900913 😄 ; not sure what the differences are between that and EPSG:3857).

**Reply by tbergama** [reply]
That's 9r8, very cheeky Google 😄 You're right, the options you shared are much lighter weight I'll definitely switch to one of those!
